docs: remove activatedRouteData backward accesses in aio guides

dario-piotrowicz opened the pull request:
developers should not access the router-outlet directive in their
template before defining a template variable for it, such
implementation is present in a couple of aio guides, fix such guides
so that they show the more correct way of accessing the outlet's data

resolves #36173

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #36173


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
 @atscott :slightly_smiling_face::+1:

dario-piotrowicz committed:
docs: remove activatedRouteData backward accesses in aio guides

developers should not access the router-outlet directive in their
template before defining a template variable for it, such
implementation is present in a couple of aio guides, fix such guides
so that they show the more correct way of accessing the outlet's data

resolves #36173

atscott commented on aio/content/guide/route-animations.md:
````diff
@@ -59,15 +59,34 @@ In addition to `path` and `component`, the `data` property of each route defines
-Here, the `prepareRoute()` method takes the value of the outlet directive (established through `#outlet="outlet"`) and returns a string value representing the state of the animation based on the custom data of the current active route. Use this data to control which transition to execute for each route.
+Here, the `prepareRoute()` method takes the value of the outlet) and returns a string value representing the state of the animation based on the custom data of the current active route. Use this data to control which transition to execute for each route.
+
+<div class="alert is-helpful">
+
+  The `router-outlet` directive holds a `activatedRouteData` property which can be accessed by assigning a template variable to the directive and using that to access the directive's data.
+
+  Like so for example:
+  ```html
+    <div [@routeAnimations]="prepareRoute(outlet)">
+      <router-outlet #outlet="outlet"></router-outlet>
+    </div>
+  ```
+  ```typescript
+    prepareRoute(outlet: RouterOutlet) {
+      return outlet?.activatedRouteData?.['animation'];
+    }
+  ```
+
+  That is not however the correct way to access the outlet's data since, it is accessing the outlet before it is defined (as in, the `outlet` usage is before the `outlet` definition in the template), this is called __backward reference__ and can cause various issues since it is not supported by Angular.
+</div>
````
We don't generally document what _not_ to do but instead just document the correct way of doing things. There are a million different ways to do something incorrectly and we shouldn't attempt to document them.

dario-piotrowicz replied:
mh... yeah you do make a good point, I just feel that this is so common and so much used that it may be better to upfront let the user know that this is not the best way to do it :thinking: 

For example at the end of the animations guide we provide a link to a angular-connect talk ([here](https://angular.io/guide/animations#more-on-angular-animations)), even in the talk itself they show this way of accessing the outlet data: [video at 16:56](https://youtu.be/rnTK9meY5us?t=1016)

Lots of other places use/teach this way, like this article I read a while back: [indepth angular animations - route animations](https://indepth.dev/posts/1285/in-depth-guide-into-animations-in-angular#route-animations), this [video tutorial](https://youtu.be/7JA90VI9fAI?t=107) this [medium post](https://medium.com/ngconf/animating-angular-route-transitions-ef02b871cc30) and the list can go on and on I think :confused: 

So yeah if you really dislike it I can remove it, or I could also wrap the code in [`avoid` code snippets](https://angular.io/guide/docs-style-guide#displaying-bad-code)

Please let me know what you prefer :slightly_smiling_face:

dario-piotrowicz replied:
I see :slightly_smiling_face: , ok let me remove that then :slightly_smiling_face: 

About breaking changes, sorry I didn't think this was something that was ok in the past and not anymore (ah... I guess it's related on the way Ivy parses the templates? :thinking:)

I really like the angular connect video, so I wouldn't really mind keeping it there (all that is says is still relevant/current in the animations api), but yeah I agree that it's sort of presents the same info but just in video format :+1: (but if you want I can open a PR to remove it :slightly_smiling_face:)

I did search "backward reference" in the repoand nothing in the docs popps up so it's either called something different or it's not mentioned at all:
![indipendent-on-open](https://user-images.githubusercontent.com/61631103/154770569-59a8a618-5062-4a52-b38f-060f757c402b.gif)

If you want I'd like to add something in the docs around this :slightly_smiling_face:

atscott replied:
In the [v9 compatibility](https://v9.angular.io/guide/ivy-compatibility) docs, it was noted as "Forward references to directive inputs accessed through local refs are no longer supported by default." In other internal docs, we had it described as "backwards references". I know this mentions inputs specifically, but I don't think it's really limited to that (but rather, things that change during the execution of the component, which is more of a mouthful and harder to explain and document). 

I think this generally falls into the "unidirectional data flow" concept. I'm not super familiar with the entirety of our docs so I don't know if this is even mentioned specifically.

dario-piotrowicz replied:
I see, thanks, I think I will have a look around to see if I can find it somewhere, and if not I'll see if I can find a good place to add it in :slightly_smiling_face::+1:

dario-piotrowicz replied:
ah yeah, sorry I forgot about that... yeah hopefully people can see that the video is a 2017 one so may be cautious that it may be slightly out of date... :thinking: 

not sure, if you want I think we can totally remove it, I really like it as it is indeed well presented and engaging, but yeah it is indeed not ideal that it mentions that way of accessing the route data :sweat:

I am sort on the fence I think I might prefer to keep it in, but just slightly (it also inconsistent with other guides as far I can tell, not many guides present videos of conferences)

:shrug:

atscott replied:
You like it so let's keep it :)

dario-piotrowicz replied:
:smile::+1:

atscott commented on aio/content/guide/route-animations.md:
```diff
@@ -59,15 +59,34 @@ In addition to `path` and `component`, the `data` property of each route defines
 After configuring the routes, add a `<router-outlet>` inside the root `AppComponent` template. The `<router-outlet>` directive tells the Angular router where to render the views when matched with a route.
 
-The `<router-outlet>` directive holds the custom data set for the currently active route which can be accessed via the directive's `activatedRouteData` property, we can use such data to animate our routing transitions.
+The `<router-outlet>` directive holds the custom data set for the currently active route which can be accessed via the route's `activatedRouteData` property, we can use such data to animate our routing transitions.
```
```suggestion
The `ChildrenOutletContexts` holds information about outlets and activated routes. We can use the `data` property of each `Route` to animate our routing transitions.
```

atscott reviewed (changes requested):
This looks pretty good! Thanks for making this update!

dario-piotrowicz committed:
fixup! docs: remove activatedRouteData backward accesses in aio guides

dario-piotrowicz commented on aio/content/guide/router-tutorial-toh.md:
```diff
@@ -1225,8 +1225,8 @@ This example uses a variable of `routerOutlet`.
 
-The `@routeAnimation` property is bound to the `getAnimationData()` with the provided `routerOutlet` reference, so the next step is to define that function in the `AppComponent`.
-The `getAnimationData()` function returns the animation property from the `data` provided through the `ActivatedRoute`. The `animation` property matches the `transition` names you used in the `slideInAnimation` defined in `animations.ts`.
+The `@routeAnimation` property is bound to the `getAnimationData()` which returns the animation property from the `data` provided by the primary route. The `animation` property matches the `transition` names you used in the `slideInAnimation` defined in `animations.ts`.
+
```
@atscott I am slightly worries that this could be a bit intimidating for new comers in the ToH tutorial :confused:

Am I overthinking it?

atscott replied:
I wouldn't be _too_ worried about it. This tutorial is and was already quite complicated. Animations and router alone are hard to understand and this includes both of them. Even the previous example used a local template ref and a variable read of a very specific property on the `RouterOutlet` which would still require understanding of how `RouterOutlet` works and what an `ActivatedRoute` is. You're right that this is certainly a step beyond that in how confusing it is, but the unfortunate truth is that it's really the only fully correct way of doing this.

dario-piotrowicz replied:
ok thanks a lot for your input :slightly_smiling_face:

atscott replied:
Oh, one more kind of cool side effect of this new tutorial is that it's actually _more_ powerful in a way. The `router-outlet` can now be defined in an entirely different template because we don't need a local ref to it anymore. So you can do route animations in entirely separate templates!

dario-piotrowicz replied:
ah wow yeah I hadn't thought of that! :thinking:  :star_struck::+1:

dario-piotrowicz commented:
@atscott thanks a lot for the lightning fast review! (the PR was still even in draft :stuck_out_tongue:)

I addressed your comments (except one to which I commented) and left you a small question in the toh section :slightly_smiling_face:

dario-piotrowicz committed:
fixup! docs: remove activatedRouteData backward accesses in aio guides

dario-piotrowicz committed:
fixup! docs: remove activatedRouteData backward accesses in aio guides

dario-piotrowicz committed:
fixup! docs: remove activatedRouteData backward accesses in aio guides

dario-piotrowicz commented:
@atscott all comments addressed plus a couple of small extra tweaks :slightly_smiling_face::+1:

jessicajaniuk reviewed (approved):
LGTM :cookie:

jessicajaniuk commented:
This PR was merged into the repository by commit bad509688b7a4c5e39d2d912abacbf5cfe2b475b.
